File: lab5/lab5_2.cpp
#include <iostream>
#include <string>

using namespace std;

class Employee {
public:
  Employee() {
    name = "Unknown";
    employeeId = 0;
  }

  Employee(const string& n, int id) {
    name = n;
    employeeId = id;
  }

  void displayNameId() const {
    cout << "ID: " << employeeId << ", Name: " << name;
  }

  void displayRole() const {
    cout << ", Role: Employee";
  }

  void displayBonus() const {
    cout << ", Bonus: $100";
  }

  void displayInfo() const {
    displayNameId();
    displayRole();
    displayBonus();
    cout << '\n';
  }

private:
  string name;
  int employeeId;
};

class Manager : public Employee {
public:
  Manager() : Employee() {
    department = "Unassigned";
  }

  Manager(const string& n, int id, const string& d) : Employee(n, id) {
    department = d;
  }

  void displayDepartment() const {
    cout << ", Department: " << department;
  }

  void displayRole() const {
    cout << ", Role: Manager";
  }

  void displayBonus() const {
    cout << ", Bonus: $500";
  }

  void displayInfo() const {
    displayNameId();
    displayRole();
    displayDepartment();
    displayBonus();
    cout << '\n';
  }

private:
  string department;
};

class Director : public Manager {
public:
  Director() : Manager() {
    budgetAllowance = 0.0;
  }

  Director(const string& n, int id, const string& d, double b) : Manager(n, id, d) {
    budgetAllowance = b;
  }

  void displayBudget() const {
    cout << ", Budget: $" << budgetAllowance;
  }

  void displayRole() const {
    cout << ", Role: Director";
  }

  void displayBonus() const {
    cout << ", Bonus: $2000";
  }

  void displayInfo() const {
    displayNameId();
    displayRole();
    displayDepartment();
    displayBudget();
    displayBonus();
    cout << '\n';
  }

private:
  double budgetAllowance;
};

int main() {
  Employee emp("Victor Zanoaga", 101);
  emp.displayInfo();

  Manager mgr("Marcel Proust", 202, "Writing");
  mgr.displayInfo();

  Director dir("Dwayne 'The Rock' Johnson", 303, "Entertainment", 500000.0);
  dir.displayInfo();

  Director dirDefault;
  dirDefault.displayInfo();

  return 0;
}

File: lab5/lab5_2.kt
open class Employee(
  val name: String = "Unknown",
  val employeeId: Int = 0
) {
  fun displayNameId() {
    print("ID: $employeeId, Name: $name")
  }

  open fun displayRole() {
    print(", Role: Employee")
  }

  open fun displayBonus() {
    print(", Bonus: $100")
  }

  open fun displayInfo() {
    displayNameId()
    displayRole()
    displayBonus()
    println()
  }
}

open class Manager(
  name: String = "Unknown",
  employeeId: Int = 0,
  val department: String = "Unassigned"
) : Employee(name, employeeId) {

  fun displayDepartment() {
    print(", Department: $department")
  }

  override fun displayRole() {
    print(", Role: Manager")
  }

  override fun displayBonus() {
    print(", Bonus: $500")
  }

  override fun displayInfo() {
    displayNameId()
    displayRole()
    displayDepartment()
    displayBonus()
    println()
  }
}

class Director(
  name: String = "Unknown",
  employeeId: Int = 0,
  department: String = "Unassigned",
  val budgetAllowance: Double = 0.0
) : Manager(name, employeeId, department) {

  fun displayBudget() {
    print(", Budget: $${budgetAllowance}")
  }

  override fun displayRole() {
    print(", Role: Director")
  }

  override fun displayBonus() {
    print(", Bonus: $2000")
  }

  override fun displayInfo() {
    displayNameId()
    displayRole()
    displayDepartment()
    displayBudget()
    displayBonus()
    println()
  }
}

fun main() {
  val emp = Employee("Victor Zanoaga", 101)
  emp.displayInfo()

  val mgr = Manager("Marcel Proust", 202, "Writing")
  mgr.displayInfo()

  val dir = Director("Dwayne 'The Rock' Johnson", 303, "Entertainment", 500000.0)
  dir.displayInfo()

  val dirDefault = Director()
  dirDefault.displayInfo()
}